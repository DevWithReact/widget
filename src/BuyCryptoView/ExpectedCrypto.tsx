import React from 'react'
import styles from './styles.module.css'

type ExpectedCryptoType = {
    amount: number,
    denom: string,
    className?: string
}

const ExpectedCrypto: React.FC<ExpectedCryptoType> = (props) => {
    const { amount, denom, className } = props
    return (
        <div className={`${styles['expected-crypto']} ${className}`}>
            <span className={styles['expected-crypto__amount']}>{`${Math.round(amount * 1e8) / 1e8} ${denom}`}</span>
            <span className={styles['expected-crypto__info']}>Crypto you get (estimation)</span>
        </div>
    )
}

export default ExpectedCrypto