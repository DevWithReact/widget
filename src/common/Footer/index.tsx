import React from 'react'
import styles from './styles.module.css'

import LogoOnramper from '../../icons/logo.svg'

const Footer = () => {
    return (
        <footer className={styles.footer}>
            <span>Powered by</span>
            <img src={LogoOnramper} alt="logo"></img>
        </footer>
    )
}

export default Footer
