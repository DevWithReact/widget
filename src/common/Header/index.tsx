import React, { useContext } from 'react'
import styles from './styles.module.css'

import IconMenu from '../../icons/menu.svg'
import IconLeftArrow from '../../icons/arrow-left.svg'
import IconClose from '../../icons/close.svg'

import { NavContext } from '../../NavContext'
import Menu from './Menu'

type HeaderType = {
    title: string;
    backButton?: boolean
    onMenuClick?: () => void
}

const Header: React.FC<HeaderType> = (props) => {
    const { nextScreen, backScreen } = useContext(NavContext)
    const { title, backButton, onMenuClick = () => nextScreen(<Menu />) } = props

    return (
        <nav className={styles.header}>
            {backButton ? <img onClick={() => backScreen()} className={`${styles['header__child']} ${styles['header__back-icon']}`} alt="Back" src={IconLeftArrow} /> : null}
            <h1 className={`${styles['header__child']}`}>{title}</h1>
            <img onClick={onMenuClick} alt="menu" className={`${styles['header__child']} ${styles['header__burger-icon']}`} src={title === 'Menu' ? IconClose : IconMenu} />
        </nav>
    )
}

Header.defaultProps = {
    backButton: false
}

export default Header