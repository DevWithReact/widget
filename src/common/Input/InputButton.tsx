import React from 'react'
import styles from './styles.module.css'

import IconChevronRight from '../../icons/chevron_right.svg'

type InputButtonType = {
    selectedOption: string,
    label?: string,
    icon?: string,
    className?: string,
    iconPosition?: 'start' | 'end',
    onClick?: () => void
}

const InputButton: React.FC<InputButtonType> = (props) => {
    const { selectedOption, label, icon, className, iconPosition } = props
    return (
        <div onClick={props.onClick} className={`${styles['input']} ${className}`}>
            {label ? <label>{label}</label> : null}
            <div className={`${styles['input__type']} ${styles['input__type--selector']}`}>
                {icon ? <img alt="Icon" src={icon} className={`${styles['input__type__child']} ${styles.input__icon} ${iconPosition === 'end' ? styles['input__type__child--old-first'] : ''}`} /> : null}
                <span style={{ 'order': iconPosition === 'end' ? -1 : 'unset' }} className={`${styles['input__type__child']} ${iconPosition === 'end' ? styles['input__type__child--new-first'] : ''}`}>{selectedOption}</span>
                <img alt="Chevron right" src={IconChevronRight} className={`${styles['input__type__child']} ${styles.input__icon} ${styles['input__icon--chevron']}`} />
            </div >
        </div >
    )
}

InputButton.defaultProps = {
    className: '',
    iconPosition: 'start'
}

export default InputButton
