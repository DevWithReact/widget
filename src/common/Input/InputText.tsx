import React from 'react'
import styles from './styles.module.css'

function InputText(props: { disabled?: boolean, symbol?: string, placeholder?: string, label?: string, className?: string, icon?: string, iconPosition?: 'start' | 'end' }) {
    const { symbol, label = '\u00A0', className = '', icon, iconPosition = 'start', disabled = false } = props
    const placeholder = disabled ? '' : props.placeholder
    return (
        <div className={`${styles['input']} ${className}`}>
            {label ? <label>{label}</label> : null}
            <div className={`${styles['input__type']} ${styles['input__type--number']} ${disabled ? styles['input__type--number--disabled'] : ''}`}>
                {icon ? <img alt="Icon" src={icon} className={`${styles['input__type__child']} ${styles.input__icon} ${iconPosition === 'end' ? styles['input__type__child--old-first'] : ''}`} /> : null}
                <span before-content={symbol} className={`${styles['input__type__child']} ${styles.symbol}  ${iconPosition === 'end' ? styles['input__type__child--new-first'] : ''}`} style={{ 'order': iconPosition === 'end' ? -1 : 'unset' }} >
                    <input type='number' min="0" placeholder={placeholder} disabled={disabled} />
                </span>
            </div>
        </div>
    )
}

export default InputText
