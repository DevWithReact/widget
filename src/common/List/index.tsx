import React, { useCallback, useState } from 'react'
import styles from './styles.module.css'
import { ListItemType, _ListItemType } from '../types'

type ListType = {
    items: ListItemType[]
    onItemClick?: (index: number, item: ListItemType) => void
    searchable?: boolean
}

const List: React.FC<ListType> = (props) => {
    const { items, searchable = false } = props
    const { onItemClick = () => null } = props

    const handleItemClick = useCallback((index: number, item: ListItemType) => {
        onItemClick(index, item)
    }, [onItemClick])

    const [query, setQuery] = useState('')

    const filterItems = useCallback((item: ListItemType) => {
        return item.name.toLowerCase().split(' ').some((substring) => substring.toLowerCase().startsWith(query))
            || item.name.toLowerCase().toLowerCase().startsWith(query)
            || item.info?.split(' ').some((substring) => substring.toLowerCase().startsWith(query))
            || item.info?.toLowerCase().startsWith(query)
            ? true : false
    }, [query])

    return (
        <div className={`${styles.list}`}>
            {searchable &&
                <input className={styles['search-box']} type='text' value={query} onChange={(e) => setQuery(e.currentTarget.value.toLowerCase())} placeholder="Search..." autoFocus />
            }
            {
                items.map((item, i) =>
                    filterItems(item) && <ListItem
                        key={i}
                        index={i}
                        name={item.name}
                        info={item.info}
                        icon={item.icon}
                        onClick={() => handleItemClick(i, item)} />
                )
            }
        </div>
    )
}

const ListItem: React.FC<_ListItemType> = (props) => {
    const { index, name, info, icon } = props
    const { onClick = () => null } = props
    return (
        <div className={`${styles['list-item']}`} onClick={() => onClick(index)}>
            {icon ? <img alt="Icon" className={`${styles['list-item__child']} ${styles['list-item__icon']}`} src={icon} /> : null}
            <div className={styles['list-item__child']}>
                <span>{name}</span>
                {info ? <span className={`${styles['list-item__info']}`}>{info}</span> : null}
            </div>
        </div>
    )
}

export default List
