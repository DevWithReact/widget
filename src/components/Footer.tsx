import React from 'react'
import styles from '../styles.module.css'

import LogoOnramper from '../icons/logo.svg'

function Footer() {
    return (
        <div className={styles.footer}>
            <span>Powered by</span>
            <img src={LogoOnramper} alt="logo"></img>
        </div>
    )
}

export default Footer
