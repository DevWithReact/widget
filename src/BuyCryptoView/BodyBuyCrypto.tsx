import React from 'react'
import stylesCommon from '../styles.module.css'

import InputButton from '../common/Input/InputButton'
import InputText from '../common/Input/InputText'
import ExpectedCrypto from './ExpectedCrypto'

import { ListItemType } from '../common/types'

type BodyBuyCryptoType = {
    onBuyCrypto: () => void,
    openPickCrypto: () => void,
    openPickCurrency: () => void,
    openPickPayment: () => void,
    selectedCrypto: ListItemType,
    selectedCurrency: ListItemType,
    selectedPaymentMethod: ListItemType,
    expectedAmount: number,
    amountValue: number
    handleInputChange: (e: React.ChangeEvent<HTMLInputElement>) => void,
}

const BodyBuyCrypto: React.FC<BodyBuyCryptoType> = (props) => {
    const { openPickCrypto, onBuyCrypto, openPickCurrency, openPickPayment } = props
    const { selectedCrypto, selectedCurrency, selectedPaymentMethod, expectedAmount, amountValue } = props
    const { handleInputChange } = props
    return (
        <main className={stylesCommon.body}>
            <InputButton onClick={openPickCrypto} className={stylesCommon['body__child']} label="I want to buy" selectedOption={selectedCrypto.name} icon={selectedCrypto.icon} />
            <div className={`${stylesCommon['body__child']} ${stylesCommon['row-fields']}`}>
                <InputText name='amount' type='number' value={amountValue} onChange={handleInputChange} className={stylesCommon['row-fields__child']} label="Amount" symbol="$" placeholder="100" />
                <InputButton onClick={openPickCurrency} className={stylesCommon['row-fields__child']} label="Currency" selectedOption={selectedCurrency.name} icon={selectedCurrency.icon} />
            </div>
            <InputButton onClick={openPickPayment} iconPosition="end" className={stylesCommon['body__child']} label="Payment method" selectedOption={selectedPaymentMethod.name} icon={selectedPaymentMethod.icon} />
            <ExpectedCrypto className={`${stylesCommon['body__child']} ${stylesCommon.grow}`} amount={expectedAmount} denom="BTC" />
            <div className={`${stylesCommon['body__child']}`}>
                <button onClick={onBuyCrypto} className={`${stylesCommon['button-action']}`}>Continue</button>
            </div>
        </main>
    )
}

BodyBuyCrypto.defaultProps = {
    selectedCrypto: {
        name: 'Loading...',
        icon: ''
    },
    selectedCurrency: {
        name: 'Loading...',
        icon: ''
    },
    selectedPaymentMethod: {
        name: 'Loading...',
        icon: ''
    },
}

export default BodyBuyCrypto