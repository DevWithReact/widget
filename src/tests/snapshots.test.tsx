import React from 'react'
import renderer from 'react-test-renderer'
import List from '../common/List'
import IconBTC from '../icons/btc.svg'
import Footer from '../common/Footer'
import Header from '../common/Header'
import InputButton from '../common/Input/InputButton'
import InputText from '../common/Input/InputText'
import ExpectedCrypto from '../BuyCryptoView/ExpectedCrypto'
import BodyBuyCrypto from '../BuyCryptoView/BodyBuyCrypto'

function testSnapshot(element: JSX.Element) {
  const component = renderer
    .create(element)
    .toJSON();
  expect(component).toMatchSnapshot();
}

const itemsListExample = [
  {
    icon: IconBTC,
    name: "BTC",
    info: "Bitcoin"
  }
];

describe("Component snapshots", () => {
  const components = new Map([
    ["Footer", <Footer />],
    ["List", <List items={itemsListExample} />],
    ["Header", <Header title="Buy crypto" />],
    ["Header with backbutton", <Header title="Buy crypto" backButton={true} />],
    ["BodyBuyCrypto", <BodyBuyCrypto onBuyCrypto={() => null} handleInputChange={() => null} openPickCrypto={() => null} openPickCurrency={() => null} openPickPayment={() => null} selectedCrypto={itemsListExample[0]} selectedCurrency={itemsListExample[0]} selectedPaymentMethod={itemsListExample[0]} />],
    ["ExpectedCrypto", <ExpectedCrypto denom="BTC" isLoading={false} />],
    ["InputButton", <InputButton label="Label" selectedOption={itemsListExample[0].name} icon={itemsListExample[0].icon} />],
    ["InputText", <InputText name='example' label="Label" />],
  ]);

  for (const [name, elem] of components) {
    test(`${name} snapshot`,
      () => testSnapshot(elem)
    )
  }
});
