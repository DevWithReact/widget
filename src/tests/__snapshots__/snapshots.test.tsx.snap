// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshots BodyBuyCrypto snapshot 1`] = `
<main
  className="body"
>
  <div
    className="input body__child"
    onClick={[Function]}
  >
    <label>
      I want to buy
    </label>
    <div
      className="input__type input__type--selector"
    >
      <img
        alt="Icon"
        className="input__type__child input__icon "
        src="btc.svg"
      />
      <span
        className="input__type__child "
        style={
          Object {
            "order": "unset",
          }
        }
      >
        BTC
      </span>
      <img
        alt="Chevron right"
        className="input__type__child input__icon input__icon--chevron"
        src="chevron_right.svg"
      />
    </div>
  </div>
  <div
    className="body__child row-fields"
  >
    <div
      className="input row-fields__child grow"
    >
      <label>
        <span
          className="label__title"
        >
          Amount
        </span>
      </label>
      <div
        className="input__type input__type--number   "
      >
        <span
          before-content=""
          className="input__type__child symbol  "
          style={
            Object {
              "order": "unset",
            }
          }
        >
          <input
            disabled={false}
            min="0"
            name="amount"
            onChange={[Function]}
            onInput={[Function]}
            placeholder="100"
            step="any"
            type="number"
            value={100}
          />
        </span>
      </div>
    </div>
    <div
      className="input row-fields__child"
      onClick={[Function]}
    >
      <label>
        Currency
      </label>
      <div
        className="input__type input__type--selector"
      >
        <img
          alt="Icon"
          className="input__type__child input__icon "
          src="btc.svg"
        />
        <span
          className="input__type__child "
          style={
            Object {
              "order": "unset",
            }
          }
        >
          BTC
        </span>
        <img
          alt="Chevron right"
          className="input__type__child input__icon input__icon--chevron"
          src="chevron_right.svg"
        />
      </div>
    </div>
  </div>
  <div
    className="input body__child"
    onClick={[Function]}
  >
    <label>
      Payment method
    </label>
    <div
      className="input__type input__type--selector"
    >
      <img
        alt="Icon"
        className="input__type__child input__icon input__type__child--old-first"
        src="btc.svg"
      />
      <span
        className="input__type__child input__type__child--new-first"
        style={
          Object {
            "order": -1,
          }
        }
      >
        BTC
      </span>
      <img
        alt="Chevron right"
        className="input__type__child input__icon input__icon--chevron"
        src="chevron_right.svg"
      />
    </div>
  </div>
  <div
    className="expected-crypto body__child grow"
  >
    <span
      className="expected-crypto__amount"
    >
      Fetching best price...
    </span>
    <span
      className="expected-crypto__info"
    >
      Crypto you get
       (estimation)
    </span>
  </div>
  <div
    className="body__child"
  >
    <button
      className="button-action "
      disabled={false}
      onClick={[Function]}
    >
      Get crypto
    </button>
  </div>
</main>
`;

exports[`Component snapshots ExpectedCrypto snapshot 1`] = `
<div
  className="expected-crypto undefined"
>
  <span
    className="expected-crypto__amount"
  >
    0 BTC
  </span>
  <span
    className="expected-crypto__info"
  >
    Crypto you get
     (estimation)
  </span>
</div>
`;

exports[`Component snapshots Footer snapshot 1`] = `
<footer
  className="footer"
>
  <span>
    Powered by
  </span>
  <img
    alt="logo"
    src="logo.svg"
  />
</footer>
`;

exports[`Component snapshots Header snapshot 1`] = `
<nav
  className="header"
>
  <h1
    className="header__child"
  >
    Buy crypto
  </h1>
  <img
    alt="menu"
    className="header__child header__burger-icon"
    onClick={[Function]}
    src="burger_menu.svg"
  />
</nav>
`;

exports[`Component snapshots Header with backbutton snapshot 1`] = `
<nav
  className="header"
>
  <img
    alt="Back"
    className="header__child header__back-icon"
    onClick={[Function]}
    src="left_arrow.svg"
  />
  <h1
    className="header__child"
  >
    Buy crypto
  </h1>
  <img
    alt="menu"
    className="header__child header__burger-icon"
    onClick={[Function]}
    src="burger_menu.svg"
  />
</nav>
`;

exports[`Component snapshots InputButton snapshot 1`] = `
<div
  className="input "
>
  <label>
    Label
  </label>
  <div
    className="input__type input__type--selector"
  >
    <img
      alt="Icon"
      className="input__type__child input__icon "
      src="btc.svg"
    />
    <span
      className="input__type__child "
      style={
        Object {
          "order": "unset",
        }
      }
    >
      BTC
    </span>
    <img
      alt="Chevron right"
      className="input__type__child input__icon input__icon--chevron"
      src="chevron_right.svg"
    />
  </div>
</div>
`;

exports[`Component snapshots InputText snapshot 1`] = `
<div
  className="input "
>
  <label>
    Label
  </label>
  <div
    className="input__type input__type--number   "
  >
    <span
      className="input__type__child symbol  "
      style={
        Object {
          "order": "unset",
        }
      }
    >
      <input
        disabled={false}
        min="0"
        name="example"
        onChange={[Function]}
        type="text"
      />
    </span>
  </div>
</div>
`;

exports[`Component snapshots List snapshot 1`] = `
<div
  className="list"
>
  <div
    className="list-item"
    onClick={[Function]}
  >
    <img
      alt="Icon"
      className="list-item__child list-item__icon"
      src="btc.svg"
    />
    <div
      className="list-item__child"
    >
      <span>
        BTC
      </span>
      <span
        className="list-item__info"
      >
        Bitcoin
      </span>
    </div>
  </div>
</div>
`;
